File: src/runtime/module.rs
use std::collections::HashMap;

use serde::de::DeserializeOwned;
use tauri::{plugin::PluginApi, AppHandle, Runtime, WebviewWindow};

use crate::error::Result;
#[cfg(target_os = "android")]
use crate::runtime::ffi::ALooper_forThread;
use crate::runtime::worklet::{BareWorklet, Looper};

pub struct BareKit<R: Runtime> {
    id: u8,
    worklets: HashMap<u8, BareWorklet<R>>,
    looper: Looper,
}

impl<R: Runtime> BareKit<R> {
    pub fn new<C: DeserializeOwned>(_app: &AppHandle<R>, _api: PluginApi<R, C>) -> Result<Self> {
        #[cfg(not(target_os = "android"))]
        let looper = Looper();
        #[cfg(target_os = "android")]
        let looper = unsafe {
            let looper = Looper(ALooper_forThread());
            looper
        };

        Ok(Self {
            id: 0,
            worklets: HashMap::new(),
            looper,
        })
    }

    pub fn invalidate(&mut self) -> Result<()> {
        for (_, worklet) in &mut self.worklets {
            worklet.terminate();
        }
        self.worklets.clear();

        Ok(())
    }

    pub fn init(
        &mut self,
        memory_limit: usize,
        assets: Option<String>,
        window: WebviewWindow<R>,
        callback_id: u32,
    ) -> Result<u8> {
        self.id += 1;

        let worklet = BareWorklet::new(self.id, memory_limit, assets, window, callback_id);
        self.worklets.insert(self.id, worklet);

        Ok(self.id)
    }

    pub fn start_file(&mut self, id: u8, filename: String, args: Vec<String>) -> Result<()> {
        if let Some(worklet) = self.worklets.get_mut(&id) {
            Ok(worklet.start_file(filename, args, &self.looper))
        } else {
            Err(format!("No worklet with ID {id}").into())
        }
    }

    pub fn start_utf8(
        &mut self,
        id: u8,
        filename: String,
        source: String,
        args: Vec<String>,
    ) -> Result<()> {
        if let Some(worklet) = self.worklets.get_mut(&id) {
            Ok(worklet.start_utf8(filename, source, args, &self.looper))
        } else {
            Err(format!("No worklet with ID {id}").into())
        }
    }

    pub fn start_bytes(
        &mut self,
        id: u8,
        filename: String,
        source: Vec<u8>,
        args: Vec<String>,
    ) -> Result<()> {
        if let Some(worklet) = self.worklets.get_mut(&id) {
            Ok(worklet.start_bytes(filename, source, args, &self.looper))
        } else {
            Err(format!("No worklet with ID {id}").into())
        }
    }

    pub fn read(&mut self, id: u8) -> Result<Vec<u8>> {
        if let Some(worklet) = self.worklets.get_mut(&id) {
            Ok(worklet.read())
        } else {
            Err(format!("No worklet with ID {id}").into())
        }
    }

    pub fn write(&mut self, id: u8, data: Vec<u8>) -> Result<i32> {
        if let Some(worklet) = self.worklets.get_mut(&id) {
            Ok(worklet.write(data))
        } else {
            Err(format!("No worklet with ID {id}").into())
        }
    }

    pub fn update(&mut self, id: u8, readable: bool, writable: bool) -> Result<()> {
        if let Some(worklet) = self.worklets.get_mut(&id) {
            Ok(worklet.update(readable, writable))
        } else {
            Err(format!("No worklet with ID {id}").into())
        }
    }

    pub fn suspend(&mut self, id: u8, linger: i32) -> Result<()> {
        if let Some(worklet) = self.worklets.get_mut(&id) {
            Ok(worklet.suspend(linger))
        } else {
            Err(format!("No worklet with ID {id}").into())
        }
    }

    pub fn resume(&mut self, id: u8) -> Result<()> {
        if let Some(worklet) = self.worklets.get_mut(&id) {
            Ok(worklet.resume())
        } else {
            Err(format!("No worklet with ID {id}").into())
        }
    }

    pub fn wakeup(&mut self, id: u8, deadline: i32) -> Result<()> {
        if let Some(worklet) = self.worklets.get_mut(&id) {
            Ok(worklet.wakeup(deadline))
        } else {
            Err(format!("No worklet with ID {id}").into())
        }
    }

    pub fn terminate(&mut self, id: u8) -> Result<()> {
        if let Some(worklet) = self.worklets.get_mut(&id) {
            Ok(worklet.terminate())
        } else {
            Err(format!("No worklet with ID {id}").into())
        }
    }
}

File: src-api/module.ts
import { invoke, transformCallback } from "@tauri-apps/api/core"
import { emit } from "@tauri-apps/api/event"
import b4a from "b4a"

declare global {
  type Platform = "android" | "ios" | "linux" | "macos" | "windows"

  interface Window {
    __TAURI_BARE_KIT_PLUGIN_INTERNALS__: {
      platform: Platform
    }
  }
}

type on_poll_callback = ((data: { readable: boolean; writable: boolean }) => void) | null

export default class NativeBareKit {
  static platform: Platform = window.__TAURI_BARE_KIT_PLUGIN_INTERNALS__.platform

  static async notify(id: number) {
    if (this.platform !== "android") emit(`bare-kit:worklet:${id}`)
  }

  static async invalidate(): Promise<void> {
    return invoke(format("bare_invalidate"))
  }

  static async init(
    memoryLimit: number,
    assets: string | null,
    pollCallback: on_poll_callback,
  ): Promise<number> {
    const onPoll = pollCallback != null ? transformCallback(pollCallback as any, false) : null
    return invoke<number>(format("bare_init"), {
      payload: { memoryLimit, assets, onPoll },
    })
  }

  static async startFile(id: number, filename: string, args: string[] = []): Promise<void> {
    return invoke(format("bare_start_file"), { payload: { id, filename, args } })
  }

  static async startUTF8(
    id: number,
    filename: string,
    source: string,
    args: string[] = [],
  ): Promise<void> {
    return invoke(format("bare_start_utf8"), { payload: { id, filename, source, args } })
  }

  static async startBytes(
    id: number,
    filename: string,
    source: Uint8Array,
    args: string[] = [],
  ): Promise<void> {
    return invoke(format("bare_start_bytes"), { payload: { id, filename, source, args } })
  }

  static async read(id: number): Promise<Uint8Array | null> {
    let data: any = await invoke<ArrayBuffer | string>(format("bare_read"), { payload: { id } })

    if (this.platform === "android") {
      data = b4a.from(data, "base64")
    }

    if (data.byteLength === 0) {
      return null
    }

    return b4a.from(data)
  }

  static async write(id: number, data: Uint8Array | null): Promise<number> {
    let payload: any = data ? b4a.concat([b4a.alloc(1, id), data]) : b4a.allocUnsafe(0)

    if (this.platform === "android") {
      payload = { payload: b4a.toString(payload, "base64") }
    }

    return invoke<number>(format("bare_write"), payload)
  }

  static async update(id: number, readable: boolean, writable: boolean): Promise<void> {
    return invoke(format("bare_update"), { payload: { id, readable, writable } })
  }

  static async suspend(id: number, linger: number): Promise<void> {
    return invoke(format("bare_suspend"), { payload: { id, linger } })
  }

  static async resume(id: number): Promise<void> {
    return invoke(format("bare_resume"), { payload: { id } })
  }

  static async wakeup(id: number, deadline: number): Promise<void> {
    return invoke(format("bare_wakeup"), { payload: { id, deadline } })
  }

  static async terminate(id: number): Promise<void> {
    return invoke(format("bare_terminate"), { payload: { id } })
  }
}

function format(fn: string): string {
  return `plugin:bare-kit|${fn}`
}
